specs/makeStyles.md

```diff
@@ -0,0 +1,322 @@
+```
+
+Each `matcher` is a function with takes a consumer-defined input and returns a boolean.
+When `useStyles` is called it calls all the matchers from top to bottom, passing its first argument to them. For every matcher which returns truthy, its styles are used.
+
+There is a special `null` matcher which always matches. We might want to change the always-matching signature to `() => true` to get rid of the special value or use a more descriptive constant like `MATCH_ALWAYS`.
+
+### Applying theme to styles
+
+No matter what theme is used, the component styles are always the same. The only way how to change the component appearance is through tokens which can be used in style values.
+The `styles` passed to `makeStyles([[matcher, styles]])` call can be either an object or a function. The function is called with all theme tokens.
+Those tokens are resolved to CSS variable usages in build time. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.
+
+::Risk:: What if we will not be able to describe the theme changes using just different values?
+::Risk:: A theme consist of hundreds or even thousands of tokens. We need to verify browsers can handle this amount of CSS variables.
+
+#### IE11 compatibility
+
+IE11 does not support CSS variables. If we decide to support IE11, we will split the styles to the items which use tokens and those which do not. The ones which do not use tokens will be processed build time, for the ones with tokens, we will fall back to runtime evaluation.
+
+### Hash-based classnames
+
+We will use a hash (MurmurHash2) of the CSS property name and the value to generate a classname. That results in deterministic (=stable) classnames. With that we can support server side rendering without issues with hydration (#3692).
+
+### Passing overrides to children using `className`
+
+To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary side effect described in `useCSS` approach above.
+Parent component passes its overrides to a child as a list of classes in `className` prop. The child, takes the input classes and references the global dictionary to get back the necessary CSS attribute names to be able to perform the merge - the merge is implemented in `ax()` function as described in #16411.
+
+#### Alternative 1: encode the CSS property name to the hash
+
+Instead of using the global dictionary side effect, we can use part of the hash in the classname to encode the CSS property:
+![](makeStyles-classname.png)
+
+Then when merging in a child, we can merge the classes by replacing previous classnames with the same CSS property hash.
+This approach makes the classnames longer and according to the test does not make the merging any faster.
+
+#### Alternative 2: use special `styles` prop for overrides
```

Comment:
what if components exposed style props like `color` but those props could only accept global tokens defined in theme like `red-500` which we could easily expand to a css variable in theme

Reply:
@ling1726 - could you please share an example of such a prop? I've seen something similar in `Badge` design spec but I don't think we should accept color as a `prop`. One of the reasons is that is not themeable.

Reply:
Yes, badge was what drove this since there are color variations available for it. Having thought through it a bit, I would say that things like `color` props could be exposed for use in selectors rather and map those selectors to all the possible color variations.

Reply:
Discussed offline, let's implement ~4 components where we will solve this with what we have now. Then we will revisit and discuss the pattern/changes going forward.